src/Utilities/distance_to_set.jl

```diff
@@ -0,0 +1,123 @@
+# Copyright (c) 2017: Miles Lubin and contributors
+# Copyright (c) 2017: Google Inc.
+#
+# Use of this source code is governed by an MIT-style license that can be found
+# in the LICENSE.md file or at https://opensource.org/licenses/MIT.
+
+"""
+    distance_to_set(point::T, set::MOI.AbstractScalarSet) where {T}
+    distance_to_set(
+        point::AbstractVector{T},
+        set::MOI.AbstractVectorSet,
+    ) where {T}
+
+Compute an upper bound on the minimum distance between `point` and the closest
+feasible point in `set`. If `point` is in the set `set`, this function _must_
+return `zero(T)`.
+
+## Definition of distance
+
+The minimum distance is computed as:
+```math
+d(x, \\mathcal{K}) = \\min_{y \\in \\mathcal{K}} || x - y ||
+```
+where ``x`` is `point` and ``\\mathcal{K}`` is `set`. The norm is computed as:
+```math
+||x|| = \\sqrt{\\texttt{set_dot}(x, x, \\mathcal{K})}
+```
+where ``\\texttt{set_dot}`` is [`Utilities.set_dot`](@ref).
+
+In the default case, where the set does not have a specialized method for
+[`Utilities.set_dot`](@ref), the norm is equivalent to the Euclidean norm
+``||x|| = \\sqrt{\\sum x_i^2}``.
+
+## Why an upper bound?
+
+In most cases, `distance_to_set` should return the smallest upper bound, but it
+may return a larger value if the smallest upper bound is expensive to compute.
```

Comment:
Add something like "in which case, this will be specified in the docstring of the set"